pandas/core/indexes/base.py

```diff
@@ -2858,13 +2858,17 @@ def _intersection(self, other, sort=False):
             indexer = algos.unique1d(Index(rvals).get_indexer_non_unique(lvals)[0])
             indexer = indexer[indexer != -1]
 
-        result = other.take(indexer).unique()._values
+        result = other.take(indexer)
+        if not result.is_unique:
+            result = result.unique()._values
+        else:
+            result = result._values
```

Comment:
im surprised this makes a difference.  Index.unique should already be doing this check internally i guess

Reply:
@jreback suggested to check for uniqueness before calling unique() to save a bit of time.

Reply:
actually @jbrockmendel makes a very good point here. Let's repurpose this PR to actually do a uniquess check *internal* to unique (of course this may only make sense for > 1000 elements or something).